File: exo4.4.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

int main () {
    int fd[2];
    char str_fd[16];

    if (pipe (fd) == -1) {
        perror ("erreur pipe.\n");
        exit (1);
    }

    pid_t pid = fork ();
    if (pid == -1) {
        perror ("erreur fork ()\n");
        exit (1);
    }

    if (pid > 0) {
        close (fd[0]);
        char message[] = "Message via variable d'environnement.\n";
        write (fd[1], message, strlen (message) + 1);

        close (fd[1]);
        wait (NULL);
    } else {
        close (fd[1]);

        sprintf (str_fd, "%d", fd[0]);
        if (setenv ("PIPE_READ_FD", str_fd, 1) != 0) {
            perror ("setenv");
            exit (1);
        }

        char* args[] = {"./fils_env", NULL};
        execv (args[0], args);
    }

    return 0;
}

//fils

int main () {
    

    char* env_val = getenv ("PIPE_READ_FD");

    if (env_val == NULL) {
        fprintf (stderr, "Fils : Variable d'environnement non trouvee.\n");
        exit (1);
    } 

    int fd_lecture = atoi (env_val);
    char buffer[256];

    ssize_t n = read (fd_lecture, buffer, sizeof (buffer));
    if (n > 0) {
        printf ("Fils : via env a recu : %s\n", buffer);
    }

    close (fd_lecture);
    return 0;
}

File: exo4.4.py
import os
import sys

# Programme qui transmet un descripteur de tube via une variable d'environnement

def main():
    r, w = os.pipe()
    pid = os.fork()

    if pid > 0:  # Processus père
        os.close(r)
        message = "Message via variable d'environnement.\n"
        os.write(w, message.encode())
        os.close(w)
        os.wait()
    else:  # Processus fils
        os.close(w)
        # Définir la variable d'environnement
        os.environ['PIPE_READ_FD'] = str(r)
        # Exécuter le programme fils
        os.execl(sys.executable, sys.executable, __file__, "fils")

# Programme fils
def fils_programme():
    env_val = os.environ.get('PIPE_READ_FD')
    
    if env_val is None:
        print("Fils : Variable d'environnement non trouvee.", file=sys.stderr)
        sys.exit(1)
    
    fd_lecture = int(env_val)
    buffer = os.read(fd_lecture, 256)
    
    if buffer:
        print(f"Fils : via env a recu : {buffer.decode()}")
    
    os.close(fd_lecture)

if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "fils":
        # Exécution en tant que programme fils
        fils_programme()
    else:
        main()
